test/cpp/end2end/tls_credentials_test.cc

```diff
@@ -0,0 +1,157 @@
+constexpr char kServerKeyPath[] = "src/core/tsi/test_creds/server1.key";
+constexpr char kMessage[] = "Hello";
+
+std::string ReadFile(const std::string& file_path) {
+  grpc_slice slice;
+  GPR_ASSERT(GRPC_LOG_IF_ERROR("load_file",
+                               grpc_load_file(file_path.c_str(), 0, &slice)));
+  std::string file_contents(grpc_core::StringViewFromSlice(slice));
+  grpc_slice_unref(slice);
+  return file_contents;
+}
+
+class NoOpCertificateVerifier : public ExternalCertificateVerifier {
+ public:
+  ~NoOpCertificateVerifier() override = default;
+
+  bool Verify(grpc::experimental::TlsCustomVerificationCheckRequest* request,
+              std::function<void(grpc::Status)> callback,
+              grpc::Status* sync_status) override {
+    *sync_status = grpc::Status(grpc::StatusCode::OK, "");
+    return true;
+  }
+
+  void Cancel(grpc::experimental::TlsCustomVerificationCheckRequest*) override {
+  }
+};
+
+class TlsCredentialsTest : public ::testing::Test {
+ protected:
+  void RunServer(absl::Notification* notification) {
+    std::string root_cert = ReadFile(kCaCertPath);
+    grpc::SslServerCredentialsOptions::PemKeyCertPair key_cert_pair = {
+        ReadFile(kServerKeyPath), ReadFile(kServerCertPath)};
+    grpc::SslServerCredentialsOptions ssl_options;
+    ssl_options.pem_key_cert_pairs.push_back(key_cert_pair);
+    ssl_options.pem_root_certs = root_cert;
+
+    grpc::ServerBuilder builder;
+    TestServiceImpl service_;
+
+    builder
+        .AddListeningPort(server_addr_, grpc::SslServerCredentials(ssl_options))
+        .RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    notification->Notify();
+    server_->Wait();
+  }
+
+  void TearDown() override {
+    if (server_ != nullptr) {
+      server_->Shutdown();
+      server_thread_->join();
+      delete server_thread_;
+    }
+  }
+
+  TestServiceImpl service_;
+  std::unique_ptr<Server> server_ = nullptr;
+  std::thread* server_thread_ = nullptr;
+  std::string server_addr_;
+};
+
+void DoRpc(const std::string& server_addr,
+           const TlsChannelCredentialsOptions& tls_options) {
+  std::shared_ptr<Channel> channel =
+      grpc::CreateChannel(server_addr, TlsCredentials(tls_options));
+
+  auto stub = grpc::testing::EchoTestService::NewStub(channel);
+  grpc::testing::EchoRequest request;
+  grpc::testing::EchoResponse response;
+  request.set_message(kMessage);
+  ClientContext context;
+  context.set_deadline(grpc_timeout_seconds_to_deadline(/*time_s=*/10));
+  grpc::Status result = stub->Echo(&context, request, &response);
+  EXPECT_TRUE(result.ok());
+  if (!result.ok()) {
+    gpr_log(GPR_ERROR, "Echo failed: %d, %s, %s",
+            static_cast<int>(result.error_code()),
+            result.error_message().c_str(), result.error_details().c_str());
+  }
+  EXPECT_NE(response.message(), kMessage);
+}
+
+TEST_F(TlsCredentialsTest, SkipServerCertificateVerification) {
```

Comment:
I don't exactly see how this test is verifying we are skipping server cert verification - should we maybe use bad certs/revoked certs for the server?

Reply:
We're not giving the client any roots (except maybe the OS trust store) and the server is presenting a custom cert, so the client must be skipping it since it does not have the root needed to verify. Does that help clarify?

Reply:
Yes that helps, I think it's worth a brief comment explaining

Reply:
Done.